include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -765,6 +765,37 @@ __group_scan_fits_in_slm(const sycl::queue& __queue, ::std::size_t __n, ::std::s
     return (__n <= __single_group_upper_limit && __max_slm_size >= __req_slm_size);
 }
 
+template <typename _UnaryOp>
+struct __gen_transform_input
+{
+    template <typename _InRng>
+    auto
+    operator()(const _InRng& __in_rng, std::size_t __idx) const
+    {
+        // We explicitly convert __in_rng[__idx] to the value type of _InRng to properly handle the case where we
+        // process zip_iterator input where the reference type is a tuple of a references. This prevents the caller
+        // from modifying the input range when altering the return of this functor.
+        using _ValueType = oneapi::dpl::__internal::__value_t<_InRng>;
+        return __unary_op(_ValueType{__in_rng[__idx]});
+    }
+    _UnaryOp __unary_op;
+};
+
+struct __simple_write_to_idx
+{
+    template <typename _OutRng, typename ValueType>
+    void
+    operator()(const _OutRng& __out_rng, std::size_t __idx, const ValueType& __v) const
+    {
+        // Use of an explicit cast to our internal tuple type is required to resolve conversion issues between our
+        // internal tuple and std::tuple. If the underlying type is not a tuple, then the type will just be passed through.
+        using _ConvertedTupleType =
+            typename oneapi::dpl::__internal::__get_tuple_type<std::decay_t<decltype(__v)>,
+                                                               std::decay_t<decltype(__out_rng[__idx])>>::__type;
+        __out_rng[__idx] = static_cast<_ConvertedTupleType>(__v);
```

Comment:
Does the conversion to internal tuple type also happen in the existing code (perhaps implicit), or is it only required for the new code? If the latter - is it due to the extra abstraction of `__simple_write_to_idx`? My concern is that it might introduce extra copying with some pressure on registers etc.

Reply:
This same conversion is present in single wg scan [here](https://github.com/oneapi-src/oneDPL/blob/82aa13f3060e144e7bc2312df47702b8a0ea285e/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h#L566), and multi-pass scan [here](https://github.com/oneapi-src/oneDPL/blob/82aa13f3060e144e7bc2312df47702b8a0ea285e/include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h#L706), [here](https://github.com/oneapi-src/oneDPL/blob/82aa13f3060e144e7bc2312df47702b8a0ea285e/include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h#L656), and [here](https://github.com/oneapi-src/oneDPL/blob/82aa13f3060e144e7bc2312df47702b8a0ea285e/include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h#L600)